playbook/queue_consumer.go

```diff
@@ -0,0 +1,159 @@
+package playbook
+
+import (
+	"encoding/json"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/flanksource/commons/logger"
+	"github.com/flanksource/duty/models"
+	"github.com/flanksource/incident-commander/api"
+	v1 "github.com/flanksource/incident-commander/api/v1"
+	"github.com/flanksource/incident-commander/db"
+	"github.com/flanksource/incident-commander/utils"
+	"github.com/google/uuid"
+	"github.com/jackc/pgx/v5/pgxpool"
+	"gorm.io/gorm"
+)
+
+type queueConsumer struct {
+	pool                           *pgxpool.Pool
+	db                             *gorm.DB
+	tickInterval                   time.Duration
+	dbReconnectMaxDuration         time.Duration
+	dbReconnectBackoffBaseDuration time.Duration
+
+	// registry stores the list of playbook run IDs
+	// that are currently being executed.
+	registry sync.Map
+}
+
+func NewQueueConsumer(db *gorm.DB, pool *pgxpool.Pool) *queueConsumer {
+	return &queueConsumer{
+		db:                             db,
+		pool:                           pool,
+		tickInterval:                   time.Minute,
+		dbReconnectMaxDuration:         time.Minute,
+		dbReconnectBackoffBaseDuration: time.Second,
+		registry:                       sync.Map{},
+	}
+}
+
+func (t *queueConsumer) Listen() error {
+	pgNotify := make(chan string)
+	go utils.ListenToPostgresNotify(t.pool, "playbook_run_updates", t.dbReconnectMaxDuration, t.dbReconnectBackoffBaseDuration, pgNotify)
```

Comment:
I would move this method to the QueueConsumer .e.g

t.PgListen("playbook_spec_approval", approvalUpdated)